javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/StructuredConfigPropertiesBridge.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.extension;
+
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.StructuredConfigProperties;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link ConfigProperties} which resolves properties based on {@link StructuredConfigProperties}.
+ *
+ * <p>Only properties starting with "otel.instrumentation." are resolved. Others return null (or
+ * default value if provided).
+ *
+ * <p>To resolve:
+ *
+ * <ul>
+ *   <li>"otel.instrumentation" refers to the ".instrumentation.java" node
+ *   <li>The portion of the property after "otel.instrumentation." is split into segments based on
+ *       ".".
+ *   <li>For each N-1 segment, we walk down the tree to find the relevant leaf {@link
+ *       StructuredConfigProperties}.
+ *   <li>We extract the property from the resolved {@link StructuredConfigProperties} using the last
+ *       segment as the property key.
+ * </ul>
+ *
+ * <p>For example, given the following YAML, asking for {@code
+ * ConfigProperties#getString("otel.instrumentation.common.string_key")} yields "value":
+ *
+ * <pre>
+ *   instrumentation:
+ *     java:
+ *       common:
+ *         string_key: value
+ * </pre>
+ */
+final class StructuredConfigPropertiesBridge implements ConfigProperties {
```

Comment:
would make sense to instrumentation-api-incubator so that in can be used in extensions

Reply:
I wasn't sure where to include this. @open-telemetry/java-instrumentation-maintainers is this or `instrumentation-api-incubator` the right thing for this type of thing?

Reply:
javaagent modules can't be used outside the agent - please correct me if I'm wrong @laurit

Reply:
the current location seems good

> would make sense to instrumentation-api-incubator so that in can be used in extensions

@zeitlinger maybe what you really want is https://github.com/open-telemetry/opentelemetry-java/pull/6549?

Reply:
I want to use the bridge in the starter at some point here: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/428aa281d532c7f1a1953c16c26974bc997acf88/instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts#L65

But maybe that's a next step.